extensions/spark/kyuubi-spark-connector-hive/src/main/scala/org/apache/kyuubi/spark/connector/hive/ExternalCatalogCache.scala

```diff
@@ -0,0 +1,132 @@
+
+import org.apache.hadoop.conf.Configuration
+import org.apache.spark.SparkConf
+import org.apache.spark.sql.SparkSession
+import org.apache.spark.sql.hive.kyuubi.connector.HiveBridgeHelper.HiveExternalCatalog
+
+import org.apache.kyuubi.spark.connector.hive.KyuubiHiveConnectorConf.{EXTERNAL_CATALOG_POOL_SIZE, EXTERNAL_CATALOG_SHARE_POLICY}
+
+object ExternalCatalogCache {
+  private var pool: ExternalCatalogCache = _
+
+  def getOrCreate(sparkSession: SparkSession): ExternalCatalogCache = {
+    if (pool == null) {
+      synchronized {
+        if (pool == null) {
+          val conf = sparkSession.sessionState.conf
+          pool = ExternalCatalogSharePolicy.fromString(
+            conf.getConf(EXTERNAL_CATALOG_SHARE_POLICY)) match {
+            case OneForAllMode => new SingletonCatalogCache()
+            case OneForOneMode => NoopCache()
+            case OneForMultipleMode =>
+              new limitCapacityCache(conf.getConf(EXTERNAL_CATALOG_POOL_SIZE))
+          }
+        }
+      }
+    }
+    pool
+  }
+
+  private[kyuubi] def reset(): Unit = {
+    if (pool != null) {
+      pool.invalidateAll()
+      pool = null
+    }
+  }
+}
+
+abstract class ExternalCatalogCache {
+
+  def take(ticket: Ticket): HiveExternalCatalog
+
+  def invalidateAll(): Unit = {}
+}
+
+class limitCapacityCache(threshold: Int) extends ExternalCatalogCache {
+
+  private val catalogPool = new ConcurrentHashMap[String, CatalogQueue]()
+
+  override def take(ticket: Ticket): HiveExternalCatalog = {
+    val catalogQueue = catalogPool.computeIfAbsent(
+      ticket.catalogName,
+      _ => {
+        CatalogQueue(threshold, ticket)
+      })
+    catalogQueue.poll()
+  }
+
+  override def invalidateAll(): Unit = {
+    catalogPool.clear()
+  }
+}
+
+class SingletonCatalogCache() extends ExternalCatalogCache {
+  private val catalogCache = new ConcurrentHashMap[String, HiveExternalCatalog]()
+
+  override def take(ticket: Ticket): HiveExternalCatalog = {
+    catalogCache.computeIfAbsent(
+      ticket.catalogName,
+      _ => {
+        new HiveExternalCatalog(ticket.sparkConf, ticket.hadoopConf)
+      })
+  }
+}
+
+case class NoopCache() extends ExternalCatalogCache {
+
+  override def take(ticket: Ticket): HiveExternalCatalog = {
+    val externalCatalog = new HiveExternalCatalog(ticket.sparkConf, ticket.hadoopConf)
+    externalCatalog
+  }
+}
+
+case class Ticket(catalogName: String, sparkConf: SparkConf, hadoopConf: Configuration)
+
+case class CatalogQueue(threshold: Int, ticket: Ticket) {
+  private val catalogs = new ArrayBlockingQueue[HiveExternalCatalog](threshold)
+
+  def poll(): HiveExternalCatalog = synchronized {
+    if (canGrow) {
+      val catalog = new HiveExternalCatalog(ticket.sparkConf, ticket.hadoopConf)
+      assertExpectedAction(catalogs.offer(catalog))
+      catalog
+    } else {
+      val headCatalog = catalogs.remove()
+      assertExpectedAction(catalogs.offer(headCatalog))
+      headCatalog
+    }
```

Comment:
it's really hard to get this behavior and make it hard to debug if there is something wrong.. I prefer not add this cache policy.